explorer/interpreter/type_checker.cpp

```diff
@@ -4588,6 +4588,20 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
       ScopeInfo::ForClassScope(scope_info, &class_scope, std::move(bindings));
   for (Nonnull<Declaration*> m : class_decl->members()) {
     CARBON_RETURN_IF_ERROR(DeclareDeclaration(m, class_scope_info));
+
+    const auto* fun = dyn_cast<FunctionDeclaration>(m);
+    if (!fun) {
+      continue;
+    }
+
+    if (fun->virt_override() == VirtualOverride::Impl) {
+      auto vtable = (*base_class)->vtable();
+      const auto* vtable_fun = vtable[fun->name()].first;
```

Comment:
It's surprising to me that we don't need to check that the function is actually a member of a base class here. Ideally I think the check that this function does actually override a member of some base class, and the check that it has the right type to do so, should be located together. Is that possible?

Reply:
There is a switch above that ensures the function is a member of the base class, but I couldn't figure out a way to call `ExpectExactType` on the new function without calling `DeclareDeclaration` on the new function. This `DeclareDeclaration` is dependent on the derived class being completely constructed which isn't yet the case in the above switch.

Reply:
Yeah, I'm not sure either that we could group the logic of both. Seems acceptable to me.